### PR TITLE
Audiodraft

### PR DESCRIPTION
**Pull request by fstirlitz**
<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### Description of your *pull request* and other information

</details>

This adds an extractor for audiodraft.com. Supersedes #4259.

Sorry for splitting this into two commits. My regular dev machine is broken, so I uploaded this from the GitHub UI, which apparently has neither any way to amend commits, nor have multiple files modified in a single commit. (And to think multi-file changesets were half the point of git in the first place. Sigh.) Since you usually squash them anyway, I think you won’t mind.

### CODE REVIEW

File: yt_dlp/extractor/audiodraft.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            'uploader_id': data_json.get('designer_id'),
+            'webpage_url': data_json.get('entry_url'),
+            'like_count': int_or_none(data_json.get('entry_likes')),
+            'average_rating': int_or_none(data_json.get('entry_rating')),
+        }
+
+
+class AudiodraftCustomIE(AudiodraftIE):
+    IE_NAME = 'Audiodraft:custom'
+    _VALID_URL = r'https?://(?:[A-Za-z0-9\-]+)\.audiodraft\.com/entry/(?P<id>\d+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/audiodraft.py:30]:
```suggestion
    _VALID_URL = r'https?://(?:[-\w]+)\.audiodraft\.com/entry/(?P<id>\d+)'
```

File: yt_dlp/extractor/audiodraft.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    def _real_extract(self, url):
+        id = self._match_id(url)
+        webpage = self._download_webpage(url, id)
+        player_entry_id = self._search_regex(r'playAudio\(\'(player_entry_\d+)\'\);', webpage, id, 'play entry id')
+        return self._audiodraft_extract_from_id(player_entry_id)
+
+
+class AudiodraftGenericIE(AudiodraftIE):
+    IE_NAME = 'Audiodraft:generic'
+    _VALID_URL = r'https?://www\.audiodraft\.com/contests/[^/#]+\#entries&eid=(?P<id>\d+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/audiodraft.py:76]:
```suggestion
    _VALID_URL = r'https?://www\.audiodraft\.com/contests/[^/#]+#entries&eid=(?P<id>\d+)'
```

File: yt_dlp/extractor/audiodraft.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
+from .common import InfoExtractor
+from ..utils import int_or_none
+
+
+class AudiodraftIE(InfoExtractor):
```
Comment (pukkandan) [inline yt_dlp/extractor/audiodraft.py:5]: The convention is for the same to end in `BaseIE`
```suggestion
class AudiodraftBaseIE(InfoExtractor):
```

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/audiodraft.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            'uploader_id': data_json.get('designer_id'),
+            'webpage_url': data_json.get('entry_url'),
+            'like_count': int_or_none(data_json.get('entry_likes')),
+            'average_rating': int_or_none(data_json.get('entry_rating')),
+        }
+
+
+class AudiodraftCustomIE(AudiodraftBaseIE):
+    IE_NAME = 'Audiodraft:custom'
+    _VALID_URL = r'https?://(?:[-\w]+)\.audiodraft\.com/entry/(?P<id>\d+)'
```
Comment (fstirlitz) [inline yt_dlp/extractor/audiodraft.py:30]: Just to note: though it’s `https?://` here, none of the custom URLs I tried actually work via HTTPS in a browser. Who knows, maybe they will one day.

**Pull request merged**